heron/schedulers/src/java/com/twitter/heron/scheduler/mesos/framework/MesosFramework.java

```diff
@@ -0,0 +1,575 @@
+import java.util.concurrent.LinkedTransferQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+import org.apache.mesos.Protos;
+import org.apache.mesos.Scheduler;
+import org.apache.mesos.SchedulerDriver;
+
+import com.twitter.heron.spi.common.Config;
+import com.twitter.heron.spi.common.Context;
+
+/**
+ * Mesos Framework, communicating with Mesos Master, to do the actual scheduling.
+ *
+ * We put synchronized block:
+ * ```
+ * synchronized(toScheduleTasks) {...}
+ * ```
+ * in critical areas for public methods potentially modifying its value.
+ * Make sure not acquiring other locks inside this block to prevent the potential for deadlock.
+ */
+
+public class MesosFramework implements Scheduler {
+  private static final Logger LOG = Logger.getLogger(MesosFramework.class.getName());
+
+  private final Config heronConfig;
+  private final Config heronRuntime;
+
+  // ContainerIndex -> BaseContainer
+  private final Map<Integer, BaseContainer> containersInfo;
+
+  // ContainerIndex -> TaskId
+  private final Map<Integer, String> tasksId;
+
+  // Thread-safe queue of taskId, representing BaseContainer to launch
+  private final Queue<String> toScheduleTasks;
+
+  // Flag indicates whether terminated
+  private volatile boolean isTerminated;
+
+  // The Mesos' SchedulerDriver
+  private volatile SchedulerDriver driver;
+
+  // Indicates whether the MesosFramework has registered.
+  // Scheduler.registered(...) method will count down and release the latch
+  private final CountDownLatch registeredLatch;
+
+  private volatile Protos.FrameworkID frameworkId;
+
+  public MesosFramework(Config heronConfig, Config heronRuntime) {
+    this.heronConfig = heronConfig;
+    this.heronRuntime = heronRuntime;
+    this.containersInfo = new ConcurrentHashMap<>();
+    this.tasksId = new ConcurrentHashMap<>();
+    this.toScheduleTasks = new LinkedTransferQueue<>();
+
+    // Initialize it as 1. Will count down and release once got registered
+    this.registeredLatch = new CountDownLatch(1);
+    this.isTerminated = false;
+  }
+
+  // Create a topology
+  public boolean createJob(Map<Integer, BaseContainer> jobDefinition) {
+    synchronized (toScheduleTasks) {
+      if (isTerminated) {
+        LOG.severe("Job has been killed");
+        return false;
+      }
+
+      // Record the jobDefinition
+      containersInfo.putAll(jobDefinition);
+
+      // To scheduler them with 0 attempt
+      for (Map.Entry<Integer, BaseContainer> entry : jobDefinition.entrySet()) {
+        Integer containerIndex = entry.getKey();
+        BaseContainer container = entry.getValue();
+        String taskId = TaskUtils.getTaskId(container.name, 0);
+        scheduleNewTask(taskId);
+      }
+    }
+
+    return true;
+  }
+
+  // Kill a topology
+  public boolean killJob() {
+    synchronized (toScheduleTasks) {
+      if (isTerminated) {
+        LOG.info("Job has been killed");
+        return false;
+      }
+      isTerminated = true;
```

Comment:
`isTerminated` needs to be within the synchronized block; otherwise the check of isTerminated and the update of Terminated could lead to concurrent bugs. It may be a little confusing to `synchronized (toScheduleTasks)`; changed it to `synchronized (this)` to make it clear.